extensions/2.0/Khronos/KHR_materials_ior/schema/glTF.KHR_materials_ior.schema.json

```diff
@@ -0,0 +1,23 @@
+{
+    "$schema": "http://json-schema.org/draft-04/schema",
+    "title": "KHR_materials_ior glTF extension",
+    "type": "object",
+    "description": "glTF extension that defines the index of refraction of a material.",
+    "allOf": [ { "$ref": "glTFProperty.schema.json" } ],
+    "properties": {
+        "ior": {
+            "type": "number",
+            "description": "The index of refraction.",
+            "default": 1.5,
+            "minimum": 1.0,
+            "gltf_detailedDescription": "The index of refraction (IOR) is a measured physical number usually in the range between 1 and 2 that determines how much the path of light is bent, or refracted, when entering a material. It also influences the ratio between reflected and transmitted light, calculated from the Fresnel equations."
+        },
+        "iorTexture": {
+            "allOf": [ { "$ref": "textureInfo.schema.json" } ],
+            "description": "A greyscale texture that defines the index of refraction.",
+            "gltf_detailedDescription": "The texture stores the inverse of the index of refraction (1/ior) to keep the values in the range 0 to 1."
```

Comment:
Dont you mean (ior - 1) ?
If we mandate that the ior value must be >= 1 <= 2 then this should work as normalizing the value in the 0 to 1 range.
If not then I think we are wasting precision since 1 / ior values will be in the 1 - 0.5 range.
Could you please elaborate?

Reply:
Maybe I will understand if you clarify the calculations in:
dielectric_brdf =
  fresnel_mix(
    diffuse_brdf(baseColor),
    microfacet_brdf(roughness^2),
    ior)

Reply:
The ior varies between 1 (air) and up to 3 for some gemstones (see Wikipedia's [List of refractive indices](https://en.wikipedia.org/wiki/List_of_refractive_indices)).

I picked the 1/ior mapping because it covers the full range of physical IORs (1 to 3) and it is available in [Substance Painter](https://docs.substance3d.com/spdoc/creating-export-presets-98959398.html#Creatingexportpresets-Assigningchannelstooutputmaps) and supported by [Corona](https://docs.substance3d.com/integrations/vray-157352154.html) and [Vray](https://docs.substance3d.com/integrations/vray-157352154.html).

The following is a simple implementation for KHR_materials_ior:

```
fresnel_mix(base, layer, ior) = base * (1 - fr(ior)) + layer * fr(ior)
fr(ior) = f0 + (1 - f0) * (1 - cos)^5
f0 = lerp( ((1-ior)/(1+ior))^2, baseColor, metallic)
```

The notation is introduced in #1717. I was hoping to merge all PBR Next PRs together, so I used the notation in #1718 (KHR_materials_ior) and #1719 (KHR_materials_specular). In case we merge it separately, I will adapt the descriptions.

Reply:
Thanks for the clarification!
#1717 Looks like a very good suggestion.
However, I wonder if it would be possible to store F0 value in textures?  (instead of ior , or 1 / ior) 
- Realtime renderers would want to have as much pre-calculated values as possible.
This is especially true for mobile devices.

What is the upside of storing the IOR as opposed to the F0 value from your point of view?

Reply:
Storing F0 directly in the (byte) texture will have bad precision. IORs in range 1 to 3 correspond to F0 in range 0 to 0.25, which is 0 to 63 in integer.

Reply:
In that case I would propse adding a (texture F0) scale value.
This value would be used to expand the range when creating the texture and (1 / scale) when sampling the texture.
For instance a material (texture) with max F0 of 0.25 would use 4 as a multipler when creating the texture and 0.25 as a multipler when fetching the F0 from the texture.
Another material may have a max F0 of 0.1 where the multipler would be 10 (and 0.1 in the shader)
This way the overhead would be one (scalar) multiplication in the fragment shader and we could get good usage of the (limited) number of bits.

Reply:
I think a scale value works, but I am a bit afraid of the complications this will add in other areas:

Depending on the implementation, an IOR value is more useful than F0.
* Use full Fresnel equation instead of Schlick approxmation.
* Considering IOR of clearcoat layer (1.5) for base layer Fresnel.
* Refraction direction (KHR_materials_transmission, KHR_materials_volume) needs IOR.
Conversion from IOR to F0 is possible via `ior = 2 / (1-sqrt(F0))-1`.

Apart from that, I am wondering: is it really cheaper to fetch 1 uniform value (textureMaxF0) and 1 varying value (f0) and multiply it instead of fetching only 1 varying value (1/ior) and doing a bit more math to compute f0?

And my last thought on it: There are already different encodings and we will introduce yet another. I know of the following encodings:
* ior = 1/tex (V-Ray, Arnold)
* f0 = 0.08 * tex (Unreal, Cycles; limits IOR to [1,1.7])
* f0 = 0.16 * tex^2 (Frostbite, Filament; limits IOR to [1, 2.4])

We would introduce:
* f0 = textureMaxF0 * tex

Authoring tools like Substance and importers/exporters will have to handle it. Moreover, depending on how authoring tools are going to expose it in the UI, I can easily imagine that it is harder to use, artists need to know how f0 converts to IOR. Exchanging textures only without the scale factor makes them useless.

In conclusion, I think scale + f0 is perfectly fine, but only if the increased complexity in tooling and (some) implementation(s) is worth the performance gain.

Reply:
Ok, thanks for the clarification regarding usage of IOR for other features.

> Apart from that, I am wondering: is it really cheaper to fetch 1 uniform value (textureMaxF0) and 1 varying value (f0) and multiply it instead of fetching only 1 varying value (1/ior) and doing a bit more math to compute f0?

Yes I would definately say so. Especially for mobile devices. 
In a fragment shader all cycles count**
I think that making it as 'lightweight' (calculation wise) as possible is what we should aim for.

> And my last thought on it: There are already different encodings and we will introduce yet another. I know of the following encodings:
> 
> * ior = 1/tex (V-Ray, Arnold)
> * f0 = 0.08 * tex (Unreal, Cycles; limits IOR to [1,1.7])
> * f0 = 0.16 * tex^2 (Frostbite, Filament; limits IOR to [1, 2.4])

Very good point.

I guess that if we think we could increase the quality by better utilization of the bits we should consider pushing for this solution anyway. :-)

**
In some cases calculations may be hidden or shadowed by other slower operations such as texel fetches. 
Eg, in this case calculations prior to using the texel (sampler) value could be for 'free' if shader is stalling in the sampler.
Similarly, if there is another texture being sampled directly after the ior -> F0 calculation the impact could also be invisible.
However I would not like to count on it, instead try to keep calculations to a minimum.

Reply:
Thanks for the input regarding performance. I added some more details to the extension and changed the mapping from 1/ior to the one you proposed. It adds some complexity, but I hope this is justified by the better performance. A nice side-effect of this change is that with an maxF0 of 0.08 the texture becomes compatible to the Unreal and Blender `specular` parameter, decreasing complexity in the exporters/importers.

Btw., as you mentioned performance, there are actually good reasons to _not_ provide a texture for ior/f0 at all. See [this comment](https://github.com/KhronosGroup/glTF/pull/1677#issuecomment-545831193) for more details. I think that even without a texture for f0 or ior, all practical use-cases are already covered by the textures in #1719. In addtion, not offering a texture would also improve performance...

Reply:
I just realized that it is much easier to use specularFactor and specularColor of #1719 to support Unreal "specular" textures. This would mean we don't need IOR textures at all.

With #1718 and #1719, f0 and f90 are computed like this:

```
f0 = ((1-ior)/(1+ior))^2 * specularFactor * specularTexture * specularColorFactor * specularColorTexture
f90 = specularFactor * specularTexture
```

Unreal "specular" uses the following formula:

```
f0 = 0.08 * specular_unreal
f90 = 1
```

If we now set ior to 1.5 (the default; no need to use KHR_materials_ior) and specularFactor to 2, specularColorTexture is identical to the Unreal specular texture (because ((1-1.5)/(1+1.5))^2 * 2 = 0.08). Much easier than `maxF0` and `f0Texture`.

@rsahlin Would that work for you? If yes, I would remove `maxF0` and `f0Texture`. The IOR will become uniform (can be converted host-side to an f0 multiplier).

Reply:
> I just realized that it is much easier to use specularFactor and specularColor of #1719 to support Unreal "specular" textures. This would mean we don't need IOR textures at all.

This is a great idea/compromise.  I support this idea.

So specularFactor affects F90, but specularColor does not.  I am okay with that.

Q: Is this how Unreal works?  (I think that in the Autodesk Standard Surface model specularColor does affect F90.)

Reply:
> * ior = 1/tex (V-Ray, Arnold)
> * f0 = 0.08 * tex (Unreal, Cycles; limits IOR to [1,1.7])
> * f0 = 0.16 * tex^2 (Frostbite, Filament; limits IOR to [1, 2.4])

The limits are wrong, as in e.g. Blender or Unreal the input specular value
allows a range `[0 .. ∞[`. If the specular scalar is in range `[0 .. ∞[` and the specular texture is in range `[0 .. 1]`, one doese have a range of `specular * specularTexture = [0 .. ∞[`

Reply:
Also, there is no reason **not** to allow textures, as Blender or UE4 do allow them.

Reply:
> Also, there is no reason not to allow textures, as Blender or UE4 do allow them.

In Barcelona you argued one reason to not go with IOR, and instead dircet F0 specification, was because if there was a iorTexture one would have to calculate the Snell's law on a per pixel basis and it was an unnecessary cost.  Thus this idea to have a specularFactor texture and specularColor texture that is outside of the ior/Snell equation -- their work on the result of the Fresnel equations.  I haven't seen any physical situations where you want to have a varying IOR anyhow -- but maybe that could happen if one bakes together a lot of unrelated materials into one.

Reply:
Yes, I do not want to have an IOR Texture, as I do not want to have these calculations per fragment. But I do want to have a F0 Texture (and related scalar) - and exactly the same encoding like e.g. in Blender and UE4 are using and e.g. can be created with Substance Painter.
Using this encoding - for reflection - no conversion calculation is needed and the final F0 is just a multiplication of the sampled value with the scalar and a constant.
Regarding reflection, both the Metallic-Roughness and Specular-Glossiness are already using varying F0 - just with a different encoding and an additonal texture.
And this new calculated F0 can directly be passed to the current Metallic-Roughness and even (if want) Specular-Glossiness workflow functions.

see and comapre to https://academy.substance3d.com/courses/the-pbr-guide-part-2

Reply:
The specularColor texture that Tobias is proposing that would be equivalent to F0 texture btw mathematically.  Thus I think we should have ior (defines F0), specularFactor (affects both F0 and F90) and specularColor (affects F0 only.)  This achieves the flexibility you want Nopper and achieves the parameterization that I want.

Reply:
> Also, there is no reason not to allow textures, as Blender or UE4 do allow them.

Depends. Textures for IOR/F0 are ok if the material is not refractive. But as soon as a renderer supports volumes and refraction, the behavior of a 2D IOR texture that affects refraction is undefined. Yes, you can certainly implement it somehow. But for a render-independent material standard we have to be more precise.

The same reasoning btw. applies to separating IOR/F0 for specular BRDF and specular BTDF. Yes, you can somehack hack it, but it's bad for a renderer-independent standard.

> So specularFactor affects F90, but specularColor does not. I am okay with that.

Yes. Besides the compatibility to Unreal, this is also needed for backwards-compatibility to glTF Spec/Gloss material. (And we don't do anything wrong by defining it in this way, it is physically-correct: in the limit F90 is always 1 for all real-world materials). I summarized the details in [this document](https://docs.google.com/document/d/1SvXWa6XGnFgsRXa1z4xw6QR1_jMmNSUTdRAmalCgoeI/edit?usp=sharing).

Reply:
That is the reason, why we want to split up this parameter for reflection and refraction.
Also, we are not looking for an independent material standard, we are looking for glTF physically based rendering next - that's a little but important difference.
Also, even if this becomes a material standard, then a texture should already be defined. If a renderer can already handle this texture (e.g. Cycles, Eevee, UE4, ...) that is good, if some renderers can't then these companies should revisit their renderers.

Reply:
> That is the reason, why we want to split up this parameter for reflection and refraction.

Blender also _does not_ split up IOR/F0 parameters for reflection (specular BRDF) and refraction (specular BTDF) (for good reasons!). For details, see the document I linked before.

> Also, we are not looking for an independent material standard, we are looking for glTF physically based rendering next - that's a little but important difference.

I am interested in a file/material format that is not ambiguous and can be implemented in rasterizers and ray tracers, with consistent visual results.

Reply:
@UX3D-nopper I can not tell if you support ior, specularFactor, specularColor or not.

To me this proposal seems to align well with Dassault Enterprise PBR as well as Autodesk Standard Surface, it is also compatible with Unreal Engine's model, and it meets your requirement of fast.

I think that others can adopt this model if they do not already.  It makes a lot of sense.

Reply:
![image](https://user-images.githubusercontent.com/27856744/75568717-16883780-5a54-11ea-9d8e-d9de5d81983e.png)
Blender does split it up. Relection is Specular, Refraction is IOR. see also here https://docs.blender.org/manual/en/latest/render/shader_nodes/shader/principled.html

Reply:
Yes, me too, and the file and material format are two separate things. But as I said, we will discuss endless. Let's hear what the others are saying.

Reply:
Yes, there are two parameters. They behave as follows:
* "transmission" is a linear blend between a non-transparent "plastic" BSDF and a transparent "glass" BSDF.
* "specular" affects IOR/F0 of _only_ the plastic BSDF
* "IOR" affects IOR/F0 and outgoing direction of _only_ the glass BSDF

This avoids the problematic case, different IORs/F0s for outgoing direction/refraction and reflection in the glass BSDF.

Works the same as the proposed extensions #1718 and #1719 (and Enterprise PBR). #1719 gives a little bit more freedom, so that the glTF Spec/Gloss and Unity Spec/Gloss materials are also supported, and compatibility to Standard Surface is better.

Reply:
> @UX3D-nopper I can not tell if you support ior, specularFactor, specularColor or not.
> 
> To me this proposal seems to align well with Dassault Enterprise PBR as well as Autodesk Standard Surface, it is also compatible with Unreal Engine's model, and it meets your requirement of fast.
> 
> I think that others can adopt this model if they do not already. It makes a lot of sense.

Standard Surface does allow a texture. Enterprise PBR does not allow a texture. So, if we would use Standard Surface, we would not have this discussion. Only how to encode the values.

Reply:
> @UX3D-nopper I can not tell if you support ior, specularFactor, specularColor or not.
> 
> To me this proposal seems to align well with Dassault Enterprise PBR as well as Autodesk Standard Surface, it is also compatible with Unreal Engine's model, and it meets your requirement of fast.
> 
> I think that others can adopt this model if they do not already. It makes a lot of sense.

Seems, but it does not. Does not take into account, what other engines and content creation tools already solved for a good reason.

Reply:
> So, if we would use Standard Surface, we would not have this discussion. Only how to encode the values.

I think we do not need a texture for IOR.  I think it is an edge case and I've never been asked for by our artists.  Also if we did adopt a texture for it, I believe you would argue against a unified IOR here and in favor of specifying F0 directly for speed reasons (a need to calculate Snell equation on a per pixel basis.)  Thus I believe this proposal in this fashion threads the needle between a few opposing camps into a near-perfect solution.

It is okay if glTF has less degrees of freedom compared to other more complex models -- in this case an iorTexture.  Not all degrees of freedom are equal in terms of their usefulness or practicality -- and this is one of those cases in my opinion.

Reply:
![image](https://user-images.githubusercontent.com/27856744/75570916-708afc00-5a58-11ea-9120-0933ab60e6bf.png)
So your artists are not allowed to do this?

Reply:
They can plug in a monochrome texture to specularColor.  It works great.  It is more intuitively understandable as well because it just modulates the F0 value rather than a more complex non-linear response, which honestly is extremely hard to explain to artists.

Remember it works like this as already mentioned above @proog128:

```
f0 = ((1-ior)/(1+ior))^2 * specularFactor * specularTexture * specularColorFactor * specularColorTexture
f90 = specularFactor * specularTexture
```

Reply:
And it gets closer and closer to the suggestion I made.

Reply:
> @rsahlin @MiiBond Would that work for you? If yes, I would remove `maxF0` and `f0Texture`. The IOR will become uniform (can be converted host-side to an f0 multiplier).

I guess so - however we are introducing quite a few heavy texture maps. 
An RGB texture vs one channel F0 texture.
My question would be - does the added texture 'weight' give us more flexibility or would it be enough with F0?

Again, I'm asking from a performance perspective to try and keep impact down.

Reply:
> however we are introducing quite a few heavy texture maps.

`specularColorTexture` in #1719 can either be 3-channel (RGB) or 1-channel (luminance) (not yet, I have to change the description). In the end, if we combine #1718 and #1719 it is now only 2 textures (specularTexture + specularColorTexture) instead of 3 (f0Texture + specularTexture + specularColorTexture).

Reply:
Great - that answers my question and will give us a lot of flexibility and freedom if I understand it correctly :-)

Another thing we should try to solve is the different packings of the textures - for instance packing 1-channel specularColor + specularTexture with for instance the MR texture.
Ideally we want as few samplers (texture sources) as possible

Reply:
I'm marking this sub-thread resolved, because there's no `iorTexture` anymore.  Specular concerns should be discussed on the other PR.